pathfinding: probability for bimodal distribution

bitromortac opened the pull request:
This PR implements an additional probability estimator ("bimodal") based on a probability theory framework using channel capacities and previous learnings from payments as inputs. The specific assumed liquidity distribution of channel liquidity is based on research in [#5988](https://github.com/lightningnetwork/lnd/issues/5988#issuecomment-1131234858).

The computed probability consists of:
* the direct channel probability, which is estimated based on a depleted liquidity distribution model, our formulas are derived after Pickhardt et al. [arXiv:2103.08576](https://arxiv.org/abs/2103.08576)
* an extension of the probability model to incorporate knowledge decay after time for previous successes and failures
* a mixed node probability taking into account successes/failures on other channels of the node (similar to the apriori approach)

We introduce an `Estimator` interface for probability computation. The interface can also be implemented externally and passed into the server's config struct. With this PR, we have two alternative probability estimator implementations, the current default one (`apriori`) and a new opt-in one (`bimodal`). Each of those estimators have their own configuration parameters that can be supplied in `lnd.conf`. Additionally, estimators can be hot-swapped by using the functionality around `lncli setmccfg/getmccfg`.

The new estimator is expected to make somewhat smarter decisions with respect to previous learnings about payment failures and successes over channels. Performance-wise, we expect the bimodal estimator to be slower due to heavier usage of formulas (~20% slower in computational time, but larger differences may appear due to execution path dependence). The exact analysis of behavior and impact on payment reliability is still open. Users that would like to do their own comparison could use their preconditioned mission control and carry out probes for either estimator. For a fairer comparison it is important to start with equal inputs, which is why a chain of API calls restoring mission control state (`QueryMissionControl`->payment->`ResetMissionControl`-> `XImportMissionControl`) is recommended.

joostjager commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,238 @@
+	// value determines how quickly the bimodal distribution drops off from
+	// the edges of a channel. A larger value means, that the drop off is
+	// slow.
+	DefaultScaleMsat = lnwire.MilliSatoshi(300000000)
```
Isn't there a way around this constant, for example by making it relative to the channel capacity? Or would that be contrary to empirical findings?

bitromortac replied:
I haven't found a good way to get rid of it, making it capacity dependent has the tendency to favor large channels, which is why I thought keeping this value constant would be better, but I can certainly have another look.

joostjager commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,238 @@
+		// We add probabilities weighted by how recent the info is.
+		var weight float64
+		if result.SuccessAmt > 0 {
+			exponent := -float64(now.Sub(result.SuccessTime)) /
```
Why switch to decaying successes? Intuitively I always thought that assuming success until proven otherwise is a strong strategy for probability estimation.

Roasbeef replied:
Tbh, I always just thought of that as a simplifying assumption we could make to ship an initial version sooner. The intuition behind this new distribution is something like: "most of the time the capacity is in one direction or the other, either due to channels starting out like that, or the inherent asymmetric of the sources and sinks in the network (some services only recv, etc)". The approach of never decaying success sort of assumes that we're the only transactors in the network, and that the context/information that led to our past observations never changes/decays. 

As a middle ground, we could experiment with something like a piece-wise approach, where we drop (after decay) to a higher-than-baseline success probability. This would _maybe_ let us tend towards nodes that are at least able to forward regularly. Ultimately, I'm looking forward to finalizing our implementation independent path finding benchmarking framework so we can have somethign more concrete to rest our heuristics/arguments on.

bitromortac replied:
This behavior is still there, the algo will still chose successful routes until they are exhausted. I am not 100% convinced of having a symmetric behavior between successes and failures, although I tend to that. One could allow two different decay times and make it a user choice (say longer decay times for successes, shorter decay times for failures - this way we would forget about failures faster?). For now I chose to keep things minimal.

joostjager commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,238 @@
+	// The channel probability is weighted fully, all other results are
+	// weighted according to how recent the information is.
+	totalProbabilities := directProbability
+	totalWeights := 1.
```
Because the result of this function is used directly and no further mixing with the direct probability is done, it is quite different compared to how it was. Especially with large amounts of channels, it seems that the direct probability hardly has any effect anymore of the final probability estimate.

bitromortac replied:
> Especially with large amounts of channels, it seems that the direct probability hardly has any effect anymore of the final probability estimate.

I really thought hard about this aspect, it may as well be a feature. I added some extended calculations to the docstring to document behavior. I also think that this approach is not too far from the current scheme, the node weight chosen is quite conservative, I think. We can set influence of this to zero with the node weight. This way one could independently test the direct probability against the current approach.

joostjager commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,238 @@
+		}
+		if !result.SuccessTime.IsZero() {
+			// Apply time decay for the amount we can send.
+			successAmount = canSend(
```
Same comment as below about decaying successes.

bitromortac replied:
See the comment above.

joostjager reviewed (commented):
A lot of changes to the current probability estimator. It's not easy to see what the actual effect on payment success will be. Left a few comments, but for me it is hard to be certain that the change set is an improvement. 

An alternative approach would be to make a much smaller incremental step and use most of the existing algorithm with the addition of penalizing amounts close to the channel capacity harder.

bitromortac commented:
> A lot of changes to the current probability estimator. It's not easy to see what the actual effect on payment success will be. Left a few comments, but for me it is hard to be certain that the change set is an improvement.

Thank you a lot for the feedback! Just to clarify, the approach would be to provide an alternative opt-in implementation. Right, without a benchmark we won't know if there's an improvement, however, there are some limits in the math that make me confident there will be improvements (for example, if the algo learns that there's a success for a  100ksat payment, it will also favor that pair for 200k sat). In the next I'll focus on establishing a benchmark method.

> An alternative approach would be to make a much smaller incremental step and use most of the existing algorithm with the addition of penalizing amounts close to the channel capacity harder.

I was thinking about that and wanted to do that after this PR, but I agree, it would be a quicker way to benefit users if we split this PR and introduce such a penalty first. I would take the approach of the first four commits and build on top of that. Passing the capacity as a parameter is due to performance reasons (haven't compared to looking up capacity from the graph). In the long run, I think we may need to restructure mission control to provide the capacity and other graph-derived data efficiently (maintained by `ChannelRouter`).

Roasbeef commented:
> An alternative approach would be to make a much smaller incremental step and use most of the existing algorithm with the addition of penalizing amounts close to the channel capacity harder.
> I would take the approach of the first four commits and build on top of that.

Yeah FWIW doing something like that isn't mutually exclusive with this PR (can even be done concurrently with just those first few commits). In fact, the work in this PR makes adding modified estimators much easier.

bitromortac commented:
Added a suggestion for how to handle configuration. I added subgroups for the different estimators, which is a breaking change for the apriori parameters.

Shares initial commits with #6857, so if questions come up, would appreciate review comments in that PR concerning these commits.

Roasbeef commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,244 @@
+	// The channel probability is weighted fully, all other results are
+	// weighted according to how recent the information is.
+	totalProbabilities := directProbability
+	totalWeights := 1.
```
Extraneous period at the end here? Or is this somehow valid syntax for `1.0`?

Roasbeef commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,244 @@
+	return lnwire.MilliSatoshi(canSend)
+}
+
+// cannotSend returns the sendable amount over the channel, respecting time decay.
```
godoc copied from above

Roasbeef commented on routing/probability_bimodal.go:
```diff
@@ -67,27 +69,38 @@ func (p *probabilityBimodal) getLocalPairProbability(now time.Time,
 
 	// Set success and failure amounts.
-	// TODO: include time decay of information.
+
```
Extra new line, also missing new line below before the fail amount.

Roasbeef replied:
Still seems to be there (non-blocking)

Roasbeef commented on routing/probability_bimodal.go:
```diff
@@ -67,27 +69,38 @@ func (p *probabilityBimodal) getLocalPairProbability(now time.Time,
-			failAmount = result.FailAmt
+			// Apply time decay for the amount we can not send.
+			failAmount = cannotSend(
+				result.FailAmt,
```
Folding here seems to be a bit wasteful, can pack it into 2 lines or so

bitromortac replied:
Changed that :+1:

Roasbeef commented on routing/probability_bimodal.go:
```diff
@@ -67,27 +69,38 @@ func (p *probabilityBimodal) getLocalPairProbability(now time.Time,
+			// Apply time decay for the amount we can send.
+			successAmount = canSend(
+				result.SuccessAmt,
+				now,
```
Same here re folding.

Roasbeef commented on routing/probability_bimodal_test.go:
```diff
@@ -196,3 +200,55 @@ func TestSuccessProbability(t *testing.T) {
+	))
+
+	// After a long time, we want the amount to approach 0.
+	require.Equal(t, lnwire.MilliSatoshi(0), canSend(
```
👍

Roasbeef commented on lnrpc/routerrpc/routing_config.go:
```diff
@@ -8,10 +8,40 @@ import (
+	AprioriConfig *AprioriConfig `group:"apriori" namespace:"apriori" description:"configuration for the apriori pathfinding probability estimator"`
+
+	// BimodalConfig defines parameters for the bimodal probability.
+	BimodalConfig *BimodalConfig `group:"bimodal" namespace:"bimodal" description:"configuration for the bimodal pathfinding probability estimator"`
```
👍

Roasbeef reviewed (commented):
Initial pass through to catch up with the latest state of the PR, will review your comments on the tracking issue and those prior preso slides to re-build my context.

bitromortac commented on routing/probability.go:
```diff
@@ -0,0 +1,23 @@
+type Estimator interface {
+	getPairProbability(now time.Time, results NodeResults,
+		toNode route.Vertex, amt lnwire.MilliSatoshi,
+		capacity btcutil.Amount) float64
```
Outdated due to the other PR :+1:

bitromortac commented:
Rebased on https://github.com/lightningnetwork/lnd/pull/6857. Probability estimators are now swappable at run time to allow for custom A/B testing. I probably need to remove `EstimatorConfig` and just incorporate config fields directly into the probability estimator implementations, as I don't see a way around the type switches for configuration.
There are some performance issues, where a pathfinding attempt would take ~0.5s compared to ~0.1s (apriori) on my machine, which needs investigation. Probably calling `math.Exp` adds to computation time and perhaps can be replaced with a more efficient approximation.

Roasbeef commented:
> There are some performance issues, where a pathfinding attempt would take ~0.5s compared to ~0.1s (apriori) on my machine, which needs investigation. Probably calling math.Exp adds to computation time and perhaps can be replaced with a more efficient approximation.

One approach would be to make a benchmark, then expose profile capture during the benchmark. This way, we'd be able to see exactly where things are slower and if we need any memoization or w/e. You could also just try to do path finding attempts locally on the CLI, then capture a profile right before you execute the command to see where things are slowing down.

Roasbeef commented:
Can be rebased now that the dependent PR has been merged!

lightninglabs-deploy commented:
@bitromortac, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@positiveblue: review reminder
@bitromortac, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@positiveblue: review reminder
@bitromortac, remember to re-request review from reviewers when ready

Roasbeef commented on server.go:
```diff
@@ -864,7 +864,7 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 	routingConfig := routerrpc.GetRoutingConfig(cfg.SubRPCServers.RouterRPC)
 
-	estimatorCfg := routing.ProbabilityEstimatorCfg{
+	estimatorCfg := routing.AprioriConfig{
```
👍

Roasbeef commented on routing/probability_estimator.go:
```diff
@@ -0,0 +1,52 @@
+		toNode route.Vertex) float64
+
+	// config returns the current estimator's configuration.
+	config() EstimatorConfig
```
Hmm, if we want to have a stricter or even entirely distinct config here, we can use type params to abstract away from this: 
```go
type Estimator[C any] interface {
     Config() T
}
```

Thought we might want to re-examine this method here itself: if we have an interface over the estimator itself, why do we need to be able to access the config? Given I'd assume that the actual estimator gets created elsewhere, then passed in to anyplace that accepts `Estimator`.

Roasbeef replied:
Other methods that could make sense here would be something like `Purge`, `Import`, `Export`, etc. Though we don't necessarily need to fill them out in this PR.

bitromortac replied:
The issue is with `GetMissionControlConfig`, where we need to read out the config to be able to update it around the `lncli setmccfg/getmccfg` functionality. In intermediate stages of the PR I had different configs for the two estimators, but then reverted back because I thought it would be nice to be able to switch back and forth between different estimators without having to reconfigure parameters. So this way we store the entire config somewhere and can reuse it to switch, but still not convinced it's the best approach.

Roasbeef commented on routing/probability_estimator.go:
```diff
@@ -0,0 +1,52 @@
+)
+
+// Estimator estimates the probability to reach a node.
+type Estimator interface {
```
Should this just be a private interface given all the methods aren't exported? 

I think exporting the methods can make sense as a public interface, since then it would be possible to pass in ones _own_ estimator at runtime when creating lnd.

Roasbeef commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,432 @@
+// payment results based on a liquidity distribution model of the LN.
+type probabilityBimodal struct {
+	// EstimatorConfig contains configuration options for our estimator.
+	EstimatorConfig
```
So they share all the same config params so far?

Roasbeef commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,432 @@
+	results NodeResults, toNode route.Vertex) float64 {
+
+	// TODO(bitromortac): think more about this case.
+	return 1.0
```
IIRC all this needs to tell us is if the channel is online/offline or not. As we want to skip things entirely, since we made the decision to _not_ do any sort of decay for the local channels, as we could end up sort of reducing them all to zero prob, which means no payments attempted at all.

Roasbeef commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,432 @@
+	results NodeResults, toNode route.Vertex, amt lnwire.MilliSatoshi,
+	capacity lnwire.MilliSatoshi) float64 {
+
+	// Determine success and failure amounts.
```
Can be wrapped to 80 chars, few other instances here in the PR.

Roasbeef commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,432 @@
+// probability and historic forwarding data of other channels of the node we try
+// to send from.
+//
+// Goals:
```
Great documentation 👍

Roasbeef commented on lnrpc/routerrpc/router.proto:
```diff
@@ -502,6 +502,36 @@ message MissionControlConfig {
+    BimodalNodeWeight defines how strongly non-routed channels should be taken
+    into account for probability estimation. Valid values are in [0,1].
+    */
+    float BimodalNodeWeight = 7;
```
We can potentially fold these into a `oneof` that'll be populated based on the `ProbabilityModel` enum above.

Roasbeef commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -455,36 +455,95 @@ func (s *Server) GetMissionControlConfig(ctx context.Context,
+	eCfg := cfg.EstimatorConfig
+
+	// Apriori parameters.
+	resp.Config.HalfLifeSeconds = uint64(eCfg.PenaltyHalfLife.Seconds())
```
Do we need to set these still for both versions? Also what about the bimodal specific items?

Roasbeef commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -455,36 +455,95 @@ func (s *Server) GetMissionControlConfig(ctx context.Context,
 		) * time.Second,
 	}
 
+	var estimatorConfig routing.EstimatorConfig
```
Seems the TOOD in this commit message is addressed?

Roasbeef commented on routing/probability_estimator.go:
```diff
@@ -1,294 +1,84 @@
+	// getPairProbability estimates the probability of successfully
+	// traversing to toNode based on historical payment outcomes for the
+	// from node. Those outcomes are passed in via the results parameter.
+	getPairProbability(now time.Time, results NodeResults,
```
getPairProbability -> PairProbability

Roasbeef commented on routing/probability_estimator.go:
```diff
@@ -1,294 +1,84 @@
+
+	// getLocalPairProbability estimates the probability of successfully
+	// traversing our own local channels to toNode.
+	getLocalPairProbability(now time.Time, results NodeResults,
```
getLocalPairProbability -> LocalPairProbability

joostjager commented on lnrpc/routerrpc/router.proto:
```diff
@@ -502,6 +502,36 @@ message MissionControlConfig {
+    enum ProbabilityModel {
+        UNKNOWN = 0;
+        APRIORI = 1;
+        BIMODAL = 2;
```
How would things look if instead of adding all estimators into the core lnd code base, they could live in an external process that interfaces with a lean lnd instance?

Roasbeef replied:
> How would things look if instead of adding all estimators into the core lnd code base, they could live in an external process that interfaces with a lean lnd instance?

Good q, I think we're pretty much on this path now. If we take this current `Estimator` interface, then add something like a `EstimatorMemory` interface (abstraction over mission control), we can expose something like:
```
type BayesianEstimator interface {
    Estimator 
    MissionControl
}
```
 
or w/e 

This can then be exposed in the top level lnd config like some of the persistence stuff already is. If set, then lnd would just use that vs making an estimator based on the CLI args. 

This mission control cfg RPC is pretty low level, and should be considered experimental and more of an exception to the path painted above. This would just be for estimators that lnd knows of first hand. Ofc, it can also be extended to pass in estimator configs it doesn't know about directly to the estimator to hot swap.

bitromortac replied:
Will try to implement something along those lines, extracting that functionality sounds like a good approach (mission control and the probability estimator can be viewed as a logical unit). I'd only see issues in making probability queries efficient, but don't have any experience in what cross-process communication would add there.

Roasbeef replied:
> Will try to implement something along those lines, extracting that functionality sounds like a good approach

I think for the scope of this PR, it's enough just to expose the _existing_ interface from the top-level config. Then we can spin off a new issue to brainstorm what a functionally complete interface would look like.

positiveblue commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,432 @@
+	// Note that for payment amounts smaller than successAmountMsat, we get
+	// a value larger than unity, which is capped here to get a proper
+	// probability.
+	if prob < 0 {
```
q: should we log if prob < 0 / > 1 and payment amount is not smaller than `successAmountMsat`?

positiveblue reviewed (commented):
I am half done with the review, will continue tomorrow but I will leave already some comments :+1:

positiveblue commented on lnrpc/routerrpc/routing_config.go:
```diff
@@ -1,3 +1,4 @@
+//nolint:lll
```
q: does this apply to the whole file? Should we reduce the scope to the `RoutingConfg` struct?

positiveblue commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,432 @@
+
+		// We add probabilities weighted by how recent the info is.
+		var weight float64
+		if result.SuccessAmt > 0 {
```
finishing the review I saw that they can both happen :+1:

lightninglabs-deploy commented:
@bitromortac, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bitromortac, remember to re-request review from reviewers when ready

callebtc commented on routing/probability_apriori.go:
```diff
@@ -58,7 +62,8 @@ var (
 	// probability.
-	ErrInvalidHopProbability = errors.New("hop probability must be in [0;1]")
+	ErrInvalidHopProbability = errors.New("hop probability must be in " +
+		"[0;1]")
```
According to line 98, it's a closed interval which is usually written as `[x, y]`

positiveblue replied:
yep, I just wasn't use to the old way to write it `[0;1]"`

bitromortac reviewed (commented):
Thanks for reviews! I changed back to individual configs per estimator and split up the large configuration commit. I had to introduce some rewrites to do it but I think it's easier to review in this way.

In the temporary benchmark commit one can see which parts of the code take most time (mostly exponentials). If we assume 1 us to explore an edge, we'd have about 80000 * 1 us = 80 ms in the worst case to explore the whole graph, so I didn't focus yet on optimizing anything, because there may be external overhead. I'll focus on more in-depth benchmarks and tests. One preliminary finding is that with the bimodal probability we explore more edges than with apriori (no idea why), which makes the apriori approach faster in total.

Roasbeef commented on routing/missioncontrol.go:
```diff
@@ -116,9 +116,8 @@ type MissionControl struct {
-	// calculations.
-	ProbabilityEstimatorCfg
+	// AprioriConfig is the config we will use for probability calculations.
+	AprioriConfig
```
+1 for the rename

Roasbeef commented on routing/probability_apriori.go:
```diff
@@ -96,8 +102,24 @@ func (p AprioriConfig) validate() error {
+
 // AprioriEstimator returns node and pair probabilities based on historical
-// payment results.
+// payment results. It uses a preconfigured success probability value for
```
👍  re the extra docs here

Roasbeef commented on routing/probability_estimator.go:
```diff
@@ -0,0 +1,37 @@
+		toNode route.Vertex) float64
+
+	// Config returns the estimator's configuration.
+	Config() estimatorConfig
```
Only used internally, so this doesn't actually need to be public?

Roasbeef commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,513 @@
+	// DefaultBimodalHalfLife is the default value for BimodalHalfLife.
+	// We will forget about previous learnings about channel liquidity on
+	// the timescale of about a week.
+	DefaultBimodalHalfLife = time.Duration(7*24) * time.Hour
```
Interesting here re the longer timescale for forgetting compared to the apriori model, I wonder if we should also modify _that_ default value? Either way, can be tuned further once this lands and we enter the eval phase with the virtual network simulator.

Roasbeef commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,513 @@
+	// channel is taken into account.
+	BimodalNodeWeight float64
+
+	// BimodalScaleMsat describes the scale over which channels
```
solid docs here 👍

Roasbeef commented on lnrpc/routerrpc/routing_config.go:
```diff
@@ -7,28 +7,16 @@ import (
-	// results and always assume the configured a priori probability for
-	// untried connections. A value of zero will ignore the a priori
-	// probability completely and only base the probability on historical
-	// results, unless there are none available.
```
The breaking change here is worth pointing out in isolation in the release notes.

Roasbeef commented on routing/probability_estimator_test.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+// BenchmarkAprioriPairProbability benchmarks the probability calculation.
+func BenchmarkAprioriPairProbability(b *testing.B) {
```
Thx for adding this set of benchmarks! 

I think we can leave them in the tree as well, so this commit doesn't necessarily need to be temp.

Roasbeef reviewed (approved):
LGTM 🐲

joostjager commented on lnrpc/routerrpc/routing_config.go:
```diff
@@ -47,4 +35,51 @@ type RoutingConfig struct {
+	AprioriConfig *AprioriConfig `group:"apriori" namespace:"apriori" description:"configuration for the apriori pathfinding probability estimator"`
+
+	// BimodalConfig defines parameters for the bimodal probability.
+	BimodalConfig *BimodalConfig `group:"bimodal" namespace:"bimodal" description:"configuration for the bimodal pathfinding probability estimator"`
```
You could first do a commit that restructures the config for apriori only, so that the real addition of bimodal stands out more in its own commit.

bitromortac replied:
See the other comment

joostjager commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -454,55 +454,155 @@ func (s *Server) GetMissionControlConfig(ctx context.Context,
+
+	// We only populate fields based on the current estimator.
+	switch v := cfg.Estimator.Config().(type) {
+	case routing.AprioriConfig:
```
Also this commit would benefit from first making all the config restructuring changes for apriori. Maybe even in the first half of the commit stack don't even mention bimodal anywhere. And then when everything is in place, add the bimodal commits.

joostjager replied:
I would have chosen to indeed regenerate twice for the sake of getting a clean commit that adds bimodal, but of course no blocker.

joostjager commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -454,39 +454,155 @@ func (s *Server) GetMissionControlConfig(ctx context.Context,
+				)
+			}
+
+			if newCfg.Apriori.Weight != 0 {
```
This logic looks different (and breaking) from what it was previously. It now seems to use default values that can be overriden, and also weight can't be set to zero here?

positiveblue replied:
FYI: in case we need to differentiate the zero value from nil (in case the zero value is a valid one) we usually add an [extra boolean field](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/lightning.proto#L2232)

joostjager commented on server.go:
```diff
@@ -864,20 +864,58 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
-		AprioriWeight:         routingConfig.AprioriWeight,
+	// We only initialize a probability estimator if there's no custom one.
+	var estimator routing.Estimator
+	if cfg.Estimator != nil {
```
Seems this can never happen?

joostjager replied:
I mean in the code as it currently is.

joostjager commented on routing/missioncontrol.go:
```diff
@@ -212,7 +204,8 @@ type paymentResult struct {
+	log.Debugf("Instantiating mission control with config: %v, %v", cfg,
+		cfg.Estimator)
 
 	if err := cfg.validate(); err != nil {
```
The estimator has already validated the config. What's the point of doing it again?

joostjager commented on lnrpc/routerrpc/router.proto:
```diff
@@ -467,6 +467,66 @@ message SetMissionControlConfigResponse {
+    /*
+    EstimatorConfig is populated dependent on the estimator type.
+    */
+    oneof EstimatorConfig {
```
Why make it a breaking change and not follow the usual procedure of deprecating first?

Roasbeef replied:
Does any major application/wallet actually _use_ this value? We added it originally for some experiments that were never actually even deployed fwiw. I think being able to switch algos on the fly is really valuable for future tuning and experiments.

joostjager replied:
I think keeping things backwards compatible on the rpc level doesn't need to rule out the ability to switch algos on the fly. But yes, if no one uses this, there is no point in keeping the old.

positiveblue replied:
As a general rule we should never remove neither change the fields already defined in an older released version of our protos files. We should simply mark the fields (or methods/services) as [deprecated](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/lightning.proto#L4111) and add new ones for the new features.

If it is true that nobody was building on this we can live with it.

positiveblue commented on sample-lnd.conf:
```diff
@@ -1099,21 +1099,22 @@ litecoin.node=ltcd
-; values are in [0, 1]. (default: 0.5)
-; routerrpc.aprioriweight=0.3
+; The time interval with which the MC store state is flushed to the DB.
+; routerrpc.mcflushinterval=1m
```
default?

positiveblue commented on routing/pathfind.go:
```diff
@@ -46,6 +46,8 @@ type pathFinder = func(g *graphParams, r *RestrictParams,
 	[]*channeldb.CachedEdgePolicy, float64, error)
 
 var (
+	DefaultEstimator = AprioriEstimatorName
```
nit: `DefaultEstimator is the default estimator used when xyz ` (this ones are always hard for me, copilot is your friend here)

positiveblue commented on cmd/lncli/cmd_mission_control.go:
```diff
@@ -85,51 +114,126 @@ func setCfg(ctx *cli.Context) error {
+			}
+
+			if ctx.IsSet("apriorihopprob") {
+				aCfg.HopProbability = float32(ctx.Float64(
```
I guess all this conversions are because we do not have `ctx.Float32` in this library.

Should we use `double` instead of `float` in the proto or it always get cast on the other side too?

positiveblue commented on lnrpc/routerrpc/router.proto:
```diff
@@ -467,6 +467,66 @@ message SetMissionControlConfigResponse {
+        AprioriParameters apriori = 7;
+        BimodalParameters bimodal = 8;
+    }
+}
```
could we reorder this so the fields go from `1...n`? Also, any reason to skip n=3?

positiveblue reviewed (commented):
Great work @bitromortac!

We just need to fix a couple of issues and clear out a few comments to ready :rocket:

positiveblue commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,513 @@
+		}
+
+		return 1.0, nil
+	} else if prob < 0.0 {
```
thanks!

positiveblue commented on lnrpc/routerrpc/router.proto:
```diff
@@ -467,6 +467,94 @@ message SetMissionControlConfigResponse {
+    ProbabilityModel defines which probability estimator should be used in
+    pathfinding.
+    */
+    ProbabilityMonel Model = 6;
```
Does this compile? (`ProbabilityMonel`)

bitromortac replied:
No, wonder how that sneaked in :slightly_smiling_face:, also saw the issue with the `sample-lnd.conf` check. Will collect these and push later if something else comes up.

joostjager commented on lnrpc/routerrpc/router.proto:
```diff
@@ -467,6 +467,94 @@ message SetMissionControlConfigResponse {
+    uint64 minimum_failure_relax_interval = 5;
+
+    enum ProbabilityModel {
+        UNKNOWN = 0;
```
Do you need `UNKNOWN` here or can `APRIORI` be the default with value 0?

joostjager replied:
Fixed I see

joostjager commented on routing/probability_bimodal.go:
```diff
@@ -0,0 +1,536 @@
+	// DefaultBimodalDecayTime is the default value for BimodalDecayTime.
+	// We will forget about previous learnings about channel liquidity on
+	// the timescale of about a week.
+	DefaultBimodalDecayTime = time.Duration(7*24) * time.Hour
```
This is indeed spectacular

positiveblue replied:
`string("are you talking about the extra type cast")`?

joostjager replied:
Hehe, no increasing decay to a week.

bitromortac replied:
I think it makes sense to have a longer decay time. In the apriori model, we don't forget about successes, only about failures, but here failures and successes are treated symmetrically, so it would be a pity to forget about those in short time (we could make those asymmetric if we wanted to as discussed earlier).

Roasbeef replied:
Yeah I think worth it to give this a shot, it's also configurable so people can set it to other values. In retrospect maybe using a 12 hour decay (user offline for a week and has no memory) was _too_ conservative.

joostjager replied:
Related: https://lists.linuxfoundation.org/pipermail/lightning-dev/2023-February/003842.html

joostjager reviewed (commented):
Left one comment about the proto.

I don't really have an opinion about the bimodel algorithm itself. Interested to find out how you are going to evaluate whether it is better. May be hard for users to actually see any differences if they switch over.

joostjager commented:
As discussed offline, it may be worth adding clear steps to the PR description explaining how users can do a side-by-side comparison of both estimators with a "mission control state restore" in between.

Roasbeef commented:
> As discussed offline, it may be worth adding clear steps to the PR description explaining how users can do a real life side-by-side comparison of both estimators with a "mission control state restore" in between.


Sounds like a good candidate for a blog post or lnd mailing list post. Either of those would be easier to consume/distribute compared to a PR description.

joostjager commented:
I've looked up `XImportMissionControl` and traced towards `importSnapshot`:

```golang
// importSnapshot takes an existing snapshot and merges it with our current
// state if the result provided are fresher than our current results. It returns
// the number of pairs that were used.
```

Not sure if this is usable for a side-by-side comparison, because it merges mission control state rather than replacing it.

I think it is worth spending some extra time on figuring this out and documenting it somewhere to really unlock the value of this PR for users. Otherwise it might be yet another set of pathfinding config flags for which users can't really decide what to do with.

Roasbeef commented:
> Not sure if this is usable for a side-by-side comparison, because it merges mission control state rather than replacing it.

Possible to first do a reset, _then_ a merge. 

> I think it is worth spending some extra time on figuring this out and documenting it somewhere to really unlock the value of this PR for users.

I think the value is unlocked right after the merge. We can update the PR description later in a non-blocking manner. As mentioned a PR description isn't as searchable/indexable as a mailing list post or blog post.

bitromortac commented:
Pushed some minor changes mostly related to API backward compatibility. I updated the PR description to give some more context and included a hint on how one could measure differences between the two estimators (exactly, with a reset of mc in between).
